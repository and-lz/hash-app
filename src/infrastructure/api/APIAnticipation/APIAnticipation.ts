import axios, { AxiosResponse } from 'axios'
import api from '../api'

const URL = 'https://frontend-challenge-7bu3nxh76a-uc.a.run.app'

interface APIAnticipationProps {
  amount: number | string
  installments: number | string
  mdr: number | string
  days: number[]
}

interface APIAnticipationResponse {
  data: {
    [key: number]: number
  }
}

async function APIAnticipation({
  amount,
  installments,
  mdr,
  days,
}: APIAnticipationProps): Promise<AxiosResponse<APIAnticipationResponse>> {
  return await axios.post(
    URL,
    {
      amount: Number(amount),
      installments: Number(installments),
      mdr: Number(mdr),
      days,
    },
    { timeout: api.maxTimeToWaitForApisToAnswerBack },
  )
}

export default APIAnticipation
