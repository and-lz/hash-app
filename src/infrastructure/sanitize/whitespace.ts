export const removeWhitespaceInString = (value: string): string =>
  value.replace(/\s/g, '')
