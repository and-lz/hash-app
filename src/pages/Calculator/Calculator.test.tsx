import { fireEvent, render, RenderResult } from '@testing-library/react'
import APIAnticipation from 'infrastructure/api/APIAnticipation/APIAnticipation'
import { removeWhitespaceInString } from 'infrastructure/sanitization/whitespace'
import React from 'react'
import { act } from 'react-dom/test-utils'
import CalculatorPage from './Calculator'

const noConnectionModalTitle = 'Você está sem conexão'
const errorModalTitle = 'Não foi possível completar sua solicitação'
const buttonLabel = 'Consultar'

const amountLabel = 'Informe o valor da renda *'
const installmentsLabel = 'Em quantas parcelas *'
const mdrLabel = 'Informe o percentual de MDR *'

jest.mock('infrastructure/api/APIAnticipation/APIAnticipation')
const mockedAPIAnticipation = APIAnticipation as jest.MockedFunction<
  typeof APIAnticipation
>

describe('<CalculatorPage />', () => {
  let component: RenderResult
  let amountField: any
  let installmentstField: any
  let mdrField: any
  let submitButton: any
  let getByText: any
  let queryByText: any
  let getByLabelText: any
  let fillCalculatorFieldsWithValidValues: { (): void; (): void }

  beforeEach(() => {
    jest.restoreAllMocks()

    component = render(<CalculatorPage anticipationDays={[1, 15, 30, 90]} />)
    getByText = component.getByText
    queryByText = component.queryByText
    getByLabelText = component.getByLabelText
    submitButton = getByText(buttonLabel)

    amountField = getByLabelText(amountLabel)
    installmentstField = getByLabelText(installmentsLabel)
    mdrField = getByLabelText(mdrLabel)

    fillCalculatorFieldsWithValidValues = () => {
      act(() => {
        fireEvent.change(amountField, { target: { value: '300000' } })
        fireEvent.change(installmentstField, { target: { value: '12' } })
        fireEvent.change(mdrField, { target: { value: '2' } })
      })
    }
  })
  describe('Renderization', () => {
    test('Should render page with title', () => {
      getByText('Simule sua anticipação')
    })
    test('Should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
  describe('Behavior', () => {
    test('Should allow the fields to be filled', () => {
      fillCalculatorFieldsWithValidValues()
      expect(removeWhitespaceInString(amountField.value)).toBe('R$3.000,00')
      expect(installmentstField.value).toBe('12')
      expect(mdrField.value).toBe('2')
    })
    test('Should present the error modal, after a failing request to the api', async () => {
      fillCalculatorFieldsWithValidValues()

      mockedAPIAnticipation.mockRejectedValue('')

      await act(async () => {
        fireEvent.click(submitButton)
      })

      getByText(errorModalTitle)
    })
    test('Should present anticipation values, after a successful request to the api', async () => {
      fillCalculatorFieldsWithValidValues()

      mockedAPIAnticipation.mockResolvedValue({
        data: {
          // @ts-ignore
          1: 2000,
          15: 4000,
          30: 6000,
          90: 8000,
        },
      })

      await act(async () => {
        fireEvent.click(submitButton)
      })

      const [day1, day15, day30, day90] = component.queryAllByTestId(
        'result-row',
      )

      expect(day1).toHaveTextContent('Amanhã: R$ 2.000,00')
      expect(day15).toHaveTextContent('Em 15 dias: R$ 4.000,00')
      expect(day30).toHaveTextContent('Em 30 dias: R$ 6.000,00')
      expect(day90).toHaveTextContent('Em 90 dias: R$ 8.000,00')
    })
    test('Should present no connection modal when the internet is off', async () => {
      fillCalculatorFieldsWithValidValues()

      jest.spyOn(navigator, 'onLine', 'get').mockReturnValue(false)

      await act(async () => {
        fireEvent.click(submitButton)
      })

      getByText(noConnectionModalTitle)
    })
    test('Should not present no connection modal when the internet is on', async () => {
      fillCalculatorFieldsWithValidValues()

      jest.spyOn(navigator, 'onLine', 'get').mockReturnValue(true)

      await act(async () => {
        fireEvent.click(submitButton)
      })

      expect(queryByText(noConnectionModalTitle)).not.toBeInTheDocument()
    })
  })
})
