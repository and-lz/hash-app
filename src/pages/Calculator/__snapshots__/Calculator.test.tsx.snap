// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalculatorPage /> Renderization Should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="cursor: auto;"
  >
    <div>
      <div
        class="sc-ksdxAp hPGHDP"
      >
        <div
          class="sc-bdvvaa ckPhbz"
        />
        <div
          class="sc-kDThTU dOQeWD"
          width="608px"
        >
          <div
            class="sc-iqsfdx sc-crHlIS fEmQeI gDWTFd"
            width="377px"
          >
            <h1
              class="sc-hKwCoD kUMuyG"
            >
              Simule sua anticipação
            </h1>
            <span
              class="sc-dkPtyc bevlEB"
            />
            <div
              class="sc-bqiQRQ fZYwCN"
              width="251px"
            >
              <form
                class="sc-hBURRC"
              >
                <label
                  class="sc-fFehDp giHfmi"
                  for="Informe o valor da renda field"
                >
                  Informe o valor da renda
                   
                  *
                </label>
                <input
                  class="sc-fotPbf iKqKRC"
                  id="Informe o valor da renda field"
                  required=""
                  type="text"
                  value="R$ 1.000,00"
                />
                <span
                  class="sc-dkPtyc bevlEB"
                />
                <label
                  class="sc-fFehDp giHfmi"
                  for="Em quantas parcelas field"
                >
                  Em quantas parcelas
                   
                  *
                </label>
                <input
                  class="sc-fotPbf iKqKRC"
                  id="Em quantas parcelas field"
                  max="12"
                  min="1"
                  required=""
                  type="number"
                  value="12"
                />
                <div
                  class="sc-bkkfTU khHAZG"
                >
                  Máximo de 12 parcelas
                </div>
                <span
                  class="sc-dkPtyc bevlEB"
                />
                <label
                  class="sc-fFehDp giHfmi"
                  for="Informe o percentual de MDR field"
                >
                  Informe o percentual de MDR
                   
                  *
                </label>
                <input
                  class="sc-fotPbf iKqKRC"
                  id="Informe o percentual de MDR field"
                  min="0"
                  required=""
                  type="number"
                  value="1"
                />
                <span
                  class="sc-dkPtyc bevlEB"
                />
                <button
                  class="sc-gsDJrp cUUmJF"
                >
                  Consultar
                </button>
              </form>
            </div>
          </div>
          <div
            class="sc-iqsfdx sc-egiSv dvtCxf fPkldC"
            width="231px"
          >
            <h1
              class="sc-eCImvq gYYTtB"
            >
              Você receberá:
            </h1>
            <hr
              class="sc-ieebsP bWUoIY"
            />
            <span
              class="sc-dkPtyc bevlEB"
            />
            <div
              class="sc-dJjZJu gFEdfQ"
            >
              <p
                class="sc-gKckTs gJDjXS"
                data-testid="result-row"
              >
                <span>
                  Amanhã
                  :
                </span>
                 
                <br />
                <span
                  class="sc-dkPtyc gQAEQm"
                />
                <em
                  class="sc-hGPAah eHlsub"
                >
                  R$ 0,00
                </em>
              </p>
              <span
                class="sc-dkPtyc bevlEB"
              />
            </div>
            <div
              class="sc-dJjZJu gFEdfQ"
            >
              <p
                class="sc-gKckTs gJDjXS"
                data-testid="result-row"
              >
                <span>
                  Em 15 dias
                  :
                </span>
                 
                <br />
                <span
                  class="sc-dkPtyc gQAEQm"
                />
                <em
                  class="sc-hGPAah eHlsub"
                >
                  R$ 0,00
                </em>
              </p>
              <span
                class="sc-dkPtyc bevlEB"
              />
            </div>
            <div
              class="sc-dJjZJu gFEdfQ"
            >
              <p
                class="sc-gKckTs gJDjXS"
                data-testid="result-row"
              >
                <span>
                  Em 30 dias
                  :
                </span>
                 
                <br />
                <span
                  class="sc-dkPtyc gQAEQm"
                />
                <em
                  class="sc-hGPAah eHlsub"
                >
                  R$ 0,00
                </em>
              </p>
              <span
                class="sc-dkPtyc bevlEB"
              />
            </div>
            <div
              class="sc-dJjZJu gFEdfQ"
            >
              <p
                class="sc-gKckTs gJDjXS"
                data-testid="result-row"
              >
                <span>
                  Em 90 dias
                  :
                </span>
                 
                <br />
                <span
                  class="sc-dkPtyc gQAEQm"
                />
                <em
                  class="sc-hGPAah eHlsub"
                >
                  R$ 0,00
                </em>
              </p>
              <span
                class="sc-dkPtyc bevlEB"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-ksdxAp hPGHDP"
    >
      <div
        class="sc-bdvvaa ckPhbz"
      />
      <div
        class="sc-kDThTU dOQeWD"
        width="608px"
      >
        <div
          class="sc-iqsfdx sc-crHlIS fEmQeI gDWTFd"
          width="377px"
        >
          <h1
            class="sc-hKwCoD kUMuyG"
          >
            Simule sua anticipação
          </h1>
          <span
            class="sc-dkPtyc bevlEB"
          />
          <div
            class="sc-bqiQRQ fZYwCN"
            width="251px"
          >
            <form
              class="sc-hBURRC"
            >
              <label
                class="sc-fFehDp giHfmi"
                for="Informe o valor da renda field"
              >
                Informe o valor da renda
                 
                *
              </label>
              <input
                class="sc-fotPbf iKqKRC"
                id="Informe o valor da renda field"
                required=""
                type="text"
                value="R$ 1.000,00"
              />
              <span
                class="sc-dkPtyc bevlEB"
              />
              <label
                class="sc-fFehDp giHfmi"
                for="Em quantas parcelas field"
              >
                Em quantas parcelas
                 
                *
              </label>
              <input
                class="sc-fotPbf iKqKRC"
                id="Em quantas parcelas field"
                max="12"
                min="1"
                required=""
                type="number"
                value="12"
              />
              <div
                class="sc-bkkfTU khHAZG"
              >
                Máximo de 12 parcelas
              </div>
              <span
                class="sc-dkPtyc bevlEB"
              />
              <label
                class="sc-fFehDp giHfmi"
                for="Informe o percentual de MDR field"
              >
                Informe o percentual de MDR
                 
                *
              </label>
              <input
                class="sc-fotPbf iKqKRC"
                id="Informe o percentual de MDR field"
                min="0"
                required=""
                type="number"
                value="1"
              />
              <span
                class="sc-dkPtyc bevlEB"
              />
              <button
                class="sc-gsDJrp cUUmJF"
              >
                Consultar
              </button>
            </form>
          </div>
        </div>
        <div
          class="sc-iqsfdx sc-egiSv dvtCxf fPkldC"
          width="231px"
        >
          <h1
            class="sc-eCImvq gYYTtB"
          >
            Você receberá:
          </h1>
          <hr
            class="sc-ieebsP bWUoIY"
          />
          <span
            class="sc-dkPtyc bevlEB"
          />
          <div
            class="sc-dJjZJu gFEdfQ"
          >
            <p
              class="sc-gKckTs gJDjXS"
              data-testid="result-row"
            >
              <span>
                Amanhã
                :
              </span>
               
              <br />
              <span
                class="sc-dkPtyc gQAEQm"
              />
              <em
                class="sc-hGPAah eHlsub"
              >
                R$ 0,00
              </em>
            </p>
            <span
              class="sc-dkPtyc bevlEB"
            />
          </div>
          <div
            class="sc-dJjZJu gFEdfQ"
          >
            <p
              class="sc-gKckTs gJDjXS"
              data-testid="result-row"
            >
              <span>
                Em 15 dias
                :
              </span>
               
              <br />
              <span
                class="sc-dkPtyc gQAEQm"
              />
              <em
                class="sc-hGPAah eHlsub"
              >
                R$ 0,00
              </em>
            </p>
            <span
              class="sc-dkPtyc bevlEB"
            />
          </div>
          <div
            class="sc-dJjZJu gFEdfQ"
          >
            <p
              class="sc-gKckTs gJDjXS"
              data-testid="result-row"
            >
              <span>
                Em 30 dias
                :
              </span>
               
              <br />
              <span
                class="sc-dkPtyc gQAEQm"
              />
              <em
                class="sc-hGPAah eHlsub"
              >
                R$ 0,00
              </em>
            </p>
            <span
              class="sc-dkPtyc bevlEB"
            />
          </div>
          <div
            class="sc-dJjZJu gFEdfQ"
          >
            <p
              class="sc-gKckTs gJDjXS"
              data-testid="result-row"
            >
              <span>
                Em 90 dias
                :
              </span>
               
              <br />
              <span
                class="sc-dkPtyc gQAEQm"
              />
              <em
                class="sc-hGPAah eHlsub"
              >
                R$ 0,00
              </em>
            </p>
            <span
              class="sc-dkPtyc bevlEB"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
